# -*- coding: utf-8 -*-
"""Копия блокнота "AIMsets.ipynb"

Automatically generated by Colab.

Original file is located at
    https://colab.research.google.com/drive/140XWD-337-hU0TXQxxaxPkPhIYWeh0ws
"""

!wget https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/integrated_call_samples_v3.20130502.ALL.panel

!wget -nH --cut-dirs=4 -r -p -e robots=off -A 'ALL.*' https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/

"""Практическая работа. Работа с библиотекой."""

pip install hail

import hail as hl

from hail.plot import show
from pprint import pprint

"""Prepare - удаляем группу AMR (смешанные американцы) и создаем тестовую и обучающую выборки"""

!grep -E 'AFR|EUR|SAS|EAS' annotation.txt > annotation_noAMR.txt

"""STEP0 - отбираем только би-аллельные SNP и убирает инделы"""

chr1 = hl.import_vcf('ALL.chr1.phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes.vcf.gz', force_bgz=True)

#only bi-allelic SNP
chr1 = chr1.filter_rows(hl.len(chr1.alleles) == 2)
#remove indels
chr1 = chr1.filter_rows(hl.is_snp(chr1.alleles[0], chr1.alleles[1]))

chr1.show()

"""STEP1 - считаем частоты SNP"""

table = hl.import_table('annotation_noAMR.txt', impute=True).key_by('sample')

table.describe()

table.show()

chr1 = chr1.annotate_cols(pheno = table[chr1.s])

chr1.show()

chr1.col.describe()

pprint(table.aggregate(hl.agg.counter(table.super_pop)))

chr1.aggregate_cols(hl.agg.counter(chr1.pheno.super_pop))

table.describe()

chr1 = hl.variant_qc(chr1)

chr1.row.describe()

print('Samples: %d Variants: %d' % (chr1.count_cols(), chr1.count_rows()))

chr1.describe()

column = table.select("super_pop")
unique_values = column.aggregate(hl.agg.collect(column.super_pop))
pops = set(unique_values)

new_ht = table.select("super_pop")

new_ht = new_ht.annotate(frequency=hl.literal(0))

new_ht.show()